File: AWG.java
/**
 * Arrayed Waveguide Grating (AWG) - N x N wavelength demultiplexer.
 *
 * Two slab star couplers connected by an array of N_a waveguides with a
 * constant geometric path increment Delta_L. The grating order m fixes
 * Delta_L = m*lam_c/n_eff_a; the focal length f_l (Rowland circle) sets the
 * dispersion on the output face so adjacent channels are separated by
 * Delta_lam_ch.
 *
 *   Delta_L = m * lam_c / n_eff_a
 *   FSR_lam = lam_c^2 / (n_g * Delta_L)
 *   f_l     = (n_eff_s * d_aw * d_io * n_eff_a) / (m * n_g * Delta_lam_ch)
 *   theta_i = (i - (N_ch+1)/2) * d_io / f_l
 *   T_i(lam) = |sum_k A_k * exp(j*(k-c)*phi_i(lam))|^2 / (sum_k A_k)^2
 *   phi_i   = beta*Delta_L + 2*pi*n_eff_s*d_aw*theta_i/lam
 *   beta    = 2*pi*n_eff_a(lam) / lam
 *
 * Refractive indices: Sellmeier silica (Malitson 1965) + EIM (TE-TE).
 *
 * Compile: javac AWG.java
 * Run:     java AWG <W> <H> <delta%> <lam_c_um> <dlam_ch_nm> <N_ch> <N_a> <d_aw_um> <d_io_um>
 *                  [m=auto] [profile=gauss|uniform] [w_g=0.35] [span=1.0] [N=900]
 * Example: java AWG 6.0 6.0 0.75 1.55 0.8 4 30 15.0 20.0
 */
public class AWG {

    public static double nSilica(double wl) {
        double B1 = 0.6961663, B2 = 0.4079426, B3 = 0.8974794;
        double C1 = 0.0684043 * 0.0684043;
        double C2 = 0.1162414 * 0.1162414;
        double C3 = 9.896161 * 9.896161;
        double l2 = wl * wl;
        return Math.sqrt(1.0 + B1*l2/(l2-C1) + B2*l2/(l2-C2) + B3*l2/(l2-C3));
    }

    /** Returns the fundamental TE slab mode (kappa, n_eff), or null if cut off. */
    public static double[] slabFundamentalTE(double wl, double d, double nC, double nCl) {
        if (nC <= nCl) return null;
        double k0 = 2 * Math.PI / wl;
        double a = d / 2;
        double V = k0 * a * Math.sqrt(nC*nC - nCl*nCl);
        double upper = Math.min(Math.PI/2 - 1e-9, V - 1e-12);
        if (upper <= 0) return null;
        double lo = 1e-9, hi = upper;
        double flo = f(lo, V), fhi = f(hi, V);
        if (flo * fhi > 0) return null;
        for (int i = 0; i < 100; i++) {
            double mid = 0.5 * (lo + hi), fmid = f(mid, V);
            if (flo * fmid <= 0) { hi = mid; fhi = fmid; }
            else { lo = mid; flo = fmid; }
            if (hi - lo < 1e-13) break;
        }
        double u = 0.5 * (lo + hi);
        return new double[] { u/a, Math.sqrt(nC*nC - (u/a/k0)*(u/a/k0)) };
    }
    private static double f(double u, double V) {
        return u * Math.tan(u) - Math.sqrt(Math.max(V*V - u*u, 0));
    }

    /** EIM (TE-TE): returns [n_eff_a, n_eff_s] or null. */
    public static double[] eim(double wl, double W, double H, double deltaPct) {
        double nCore = nSilica(wl);
        double nClad = nCore * (1 - deltaPct / 100.0);
        double[] v = slabFundamentalTE(wl, H, nCore, nClad);
        if (v == null) return null;
        double nEffS = v[1];
        double[] h = slabFundamentalTE(wl, W, nEffS, nClad);
        if (h == null) return null;
        return new double[] { h[1], nEffS };
    }

    public static double groupIndex(double wl, double W, double H, double deltaPct) {
        double dwl = wl * 1e-4;
        double[] a = eim(wl - dwl, W, H, deltaPct);
        double[] b = eim(wl + dwl, W, H, deltaPct);
        double[] c = eim(wl, W, H, deltaPct);
        if (a == null || b == null || c == null) return Double.NaN;
        double dn = (b[0] - a[0]) / (2 * dwl);
        return c[0] - wl * dn;
    }

    public static class Design {
        double nEffA, nEffS, nG, dL, fsr, fl, sweep;
        int m;
    }

    public static Design design(double W, double H, double deltaPct, double lamC,
                                 double dlamChUm, int nCh, int nA,
                                 double dAw, double dIo, int mPref) {
        double[] at = eim(lamC, W, H, deltaPct);
        if (at == null) return null;
        double nEffA = at[0], nEffS = at[1];
        double nG = groupIndex(lamC, W, H, deltaPct);
        if (Double.isNaN(nG)) nG = nEffA;

        int m;
        if (mPref <= 0) {
            double fsrTarget = (nCh + 2) * dlamChUm;
            m = Math.max(1, (int) Math.round(lamC * nEffA / (nG * fsrTarget)));
        } else {
            m = mPref;
        }
        double dL = m * lamC / nEffA;
        double fsr = lamC * lamC / (nG * dL);
        double fl = (nEffS * dAw * dIo * nEffA) / (m * nG * dlamChUm);
        double sweep = (nCh - 1) * dIo / fl;

        Design d = new Design();
        d.nEffA = nEffA; d.nEffS = nEffS; d.nG = nG;
        d.m = m; d.dL = dL; d.fsr = fsr; d.fl = fl; d.sweep = sweep;
        return d;
    }

    public static double[] couplingAmp(int nA, String profile, double wG) {
        double[] A = new double[nA + 1];
        double c = 0.5 * (nA + 1);
        for (int k = 1; k <= nA; k++) {
            if ("gauss".equals(profile)) {
                double x = (k - c) / (wG * nA);
                A[k] = Math.exp(-x * x);
            } else {
                A[k] = 1.0;
            }
        }
        return A;
    }

    public static double transmission(double wl, int port, double W, double H, double deltaPct,
                                       double dAw, double dIo, int nA, int nCh,
                                       Design d, double[] A) {
        double[] at = eim(wl, W, H, deltaPct);
        if (at == null) return Double.NaN;
        double nEffA = at[0], nEffS = at[1];
        double beta = 2 * Math.PI * nEffA / wl;
        double cch = 0.5 * (nCh + 1);
        double theta = (port - cch) * dIo / d.fl;
        double dphi = beta * d.dL + 2 * Math.PI * nEffS * dAw * theta / wl;
        double c = 0.5 * (nA + 1);
        double re = 0, im = 0, sumA = 0;
        for (int k = 1; k <= nA; k++) {
            double phi = (k - c) * dphi;
            re += A[k] * Math.cos(phi);
            im += A[k] * Math.sin(phi);
            sumA += A[k];
        }
        return (re * re + im * im) / (sumA * sumA);
    }

    public static void main(String[] args) {
        if (args.length < 9) {
            System.out.println("Usage: java AWG <W> <H> <delta%> <lam_c_um> <dlam_ch_nm> <N_ch> <N_a>");
            System.out.println("              <d_aw_um> <d_io_um> [m=auto] [profile=gauss|uniform]");
            System.out.println("              [w_g=0.35] [span=1.0] [N=900]");
            return;
        }
        double W = Double.parseDouble(args[0]);
        double H = Double.parseDouble(args[1]);
        double deltaPct = Double.parseDouble(args[2]);
        double lamC = Double.parseDouble(args[3]);
        double dlamChNm = Double.parseDouble(args[4]);
        int nCh = Integer.parseInt(args[5]);
        int nA = Integer.parseInt(args[6]);
        double dAw = Double.parseDouble(args[7]);
        double dIo = Double.parseDouble(args[8]);
        int mPref = args.length > 9 ? Integer.parseInt(args[9]) : 0;
        String profile = args.length > 10 ? args[10] : "gauss";
        double wG = args.length > 11 ? Double.parseDouble(args[11]) : 0.35;
        double span = args.length > 12 ? Double.parseDouble(args[12]) : 1.0;
        int N = args.length > 13 ? Integer.parseInt(args[13]) : 900;

        double dlamChUm = dlamChNm * 1e-3;
        Design d = design(W, H, deltaPct, lamC, dlamChUm, nCh, nA, dAw, dIo, mPref);
        if (d == null) {
            System.out.println("No guided mode for these parameters.");
            return;
        }
        double[] A = couplingAmp(nA, profile, wG);

        System.out.printf("n_eff_a = %.6f%n", d.nEffA);
        System.out.printf("n_eff_s = %.6f%n", d.nEffS);
        System.out.printf("n_g     = %.6f%n", d.nG);
        System.out.printf("m       = %d%n", d.m);
        System.out.printf("Delta_L = %.4f um%n", d.dL);
        System.out.printf("FSR     = %.4f nm%n", d.fsr * 1000);
        System.out.printf("f_l     = %.4f mm%n", d.fl * 1e-3);
        System.out.printf("sweep   = %.3f deg%n", Math.toDegrees(d.sweep));

        System.out.println();
        StringBuilder header = new StringBuilder("# wavelength_um");
        for (int i = 1; i <= nCh; i++) header.append(String.format("  port%d_dB", i));
        System.out.println(header);
        double half = 0.5 * span * d.fsr;
        for (int n = 0; n < N; n++) {
            double wl = lamC - half + 2 * half * n / (N - 1);
            StringBuilder row = new StringBuilder(String.format("%.6f", wl));
            for (int i = 1; i <= nCh; i++) {
                double T = transmission(wl, i, W, H, deltaPct, dAw, dIo, nA, nCh, d, A);
                double dB = T > 1e-12 ? 10 * Math.log10(T) : -120.0;
                row.append(String.format("  %8.3f", dB));
            }
            System.out.println(row);
        }
    }
}

File: awg.py
"""Arrayed Waveguide Grating (AWG) - N x N wavelength demultiplexer.

Two slab star couplers connected by an array of N_a waveguides with a constant
geometric path increment Delta_L. The grating order m fixes Delta_L = m*lam_c/n_eff_a;
the focal length f_l of the Rowland-circle slab sets the linear dispersion on the
output face so adjacent channels are separated by Delta_lam_ch.

Design equations (M.K. Smit, 1996; Takahashi, 1995):
    Delta_L = m * lam_c / n_eff_a                       (path increment)
    FSR_lam = lam_c^2 / (n_g * Delta_L)                 (free spectral range)
    f_l     = (n_eff_s * d_aw * d_io * n_eff_a) / (m * n_g * Delta_lam_ch)
    theta_i = (i - (N_ch+1)/2) * d_io / f_l             (port angle)

Output transmission at port i:
    T_i(lam) = | sum_k A_k * exp(j*(k-c)*phi_i(lam)) |^2 / (sum_k A_k)^2
    phi_i(lam) = beta(lam)*Delta_L + 2*pi*n_eff_s*d_aw*theta_i / lam
    beta = 2*pi*n_eff_a(lam) / lam

A_k is the coupling profile at the slab/array junction: uniform or Gaussian
(width parameter w_g typ. 0.3-0.4 of N_a).

Refractive indices follow Sellmeier silica (Malitson 1965) + Effective Index
Method (TE-TE) for the channel waveguide; n_eff_s is the vertical slab mode
index in the star coupler.

Usage:
    python awg.py <W> <H> <delta_pct> <lam_c_um> <dlam_ch_nm> <N_ch> <N_a> <d_aw_um> <d_io_um> [m=auto] [profile=gauss|uniform] [w_g=0.35] [span=1.0] [N=900]
Example (1550 nm grid, 0.8 nm spacing, 1x4 AWG):
    python awg.py 6.0 6.0 0.75 1.55 0.8 4 30 15.0 20.0
"""

import math
import sys


def n_silica(wl):
    B1, B2, B3 = 0.6961663, 0.4079426, 0.8974794
    C1, C2, C3 = 0.0684043**2, 0.1162414**2, 9.896161**2
    l2 = wl * wl
    return math.sqrt(1 + B1*l2/(l2-C1) + B2*l2/(l2-C2) + B3*l2/(l2-C3))


def slab_te_fundamental(wl, d, n_core, n_clad):
    """Fundamental TE mode of a symmetric slab (kappa, n_eff). Returns None if cut off."""
    if n_core <= n_clad:
        return None
    k0 = 2*math.pi / wl
    a = d / 2.0
    V = k0 * a * math.sqrt(n_core*n_core - n_clad*n_clad)
    upper = min(math.pi/2 - 1e-9, V - 1e-12)
    if upper <= 0:
        return None
    def f(u):
        w = math.sqrt(max(V*V - u*u, 0.0))
        return u * math.tan(u) - w
    lo, hi = 1e-9, upper
    flo = f(lo); fhi = f(hi)
    if flo * fhi > 0:
        return None
    for _ in range(100):
        mid = 0.5 * (lo + hi)
        fmid = f(mid)
        if flo * fmid <= 0:
            hi = mid; fhi = fmid
        else:
            lo = mid; flo = fmid
        if hi - lo < 1e-13:
            break
    u = 0.5 * (lo + hi)
    return u/a, math.sqrt(n_core*n_core - (u/a/k0)**2)


def eim(wl, W, H, delta_pct):
    """Effective Index Method (TE-TE) for the channel waveguide.

    Returns (n_eff_a, n_eff_s):
        n_eff_a = channel mode index (used as beta source for the array waveguide)
        n_eff_s = vertical slab mode index (used for the star coupler region)
    """
    n_clad = n_silica(wl)
    n_core = n_clad / (1 - delta_pct/100.0)
    v = slab_te_fundamental(wl, H, n_core, n_clad)
    if v is None:
        return None
    _, n_eff_s = v
    h = slab_te_fundamental(wl, W, n_eff_s, n_clad)
    if h is None:
        return None
    _, n_eff_a = h
    return n_eff_a, n_eff_s


def group_index(wl, W, H, delta_pct):
    """n_g = n_eff_a - lam * d(n_eff_a)/d(lam) via central difference."""
    dwl = wl * 1e-4
    a = eim(wl - dwl, W, H, delta_pct)
    b = eim(wl + dwl, W, H, delta_pct)
    c = eim(wl, W, H, delta_pct)
    if a is None or b is None or c is None:
        return None
    dn_dwl = (b[0] - a[0]) / (2*dwl)
    return c[0] - wl * dn_dwl


def design(W, H, delta_pct, lam_c, dlam_ch_um, n_ch, n_a, d_aw, d_io, m_pref=0):
    """Return the AWG design parameters as a dict (all lengths in um)."""
    n_eff_a, n_eff_s = eim(lam_c, W, H, delta_pct)
    n_g = group_index(lam_c, W, H, delta_pct) or n_eff_a

    # If user did not specify the phase order, pick m such that FSR >= (N_ch+2)*Dlam_ch.
    # m = round(lam_c * n_eff_a / (n_g * FSR_target))
    if m_pref <= 0:
        fsr_target = (n_ch + 2) * dlam_ch_um
        m = max(1, round(lam_c * n_eff_a / (n_g * fsr_target)))
    else:
        m = int(m_pref)

    dL = m * lam_c / n_eff_a
    fsr = lam_c * lam_c / (n_g * dL)
    fl = (n_eff_s * d_aw * d_io * n_eff_a) / (m * n_g * dlam_ch_um)
    sweep = (n_ch - 1) * d_io / fl
    return dict(n_eff_a=n_eff_a, n_eff_s=n_eff_s, n_g=n_g,
                m=m, dL=dL, fsr=fsr, fl=fl, sweep=sweep)


def coupling_amp(n_a, profile="gauss", w_g=0.35):
    A = [0.0] * (n_a + 1)
    c = 0.5 * (n_a + 1)
    for k in range(1, n_a + 1):
        if profile == "gauss":
            x = (k - c) / (w_g * n_a)
            A[k] = math.exp(-x*x)
        else:
            A[k] = 1.0
    return A


def transmission(wl, port, p, d, A):
    """Power transmission at output `port` (1..N_ch) at wavelength wl [um]."""
    res = eim(wl, p["W"], p["H"], p["delta_pct"])
    if res is None:
        return float("nan")
    n_eff_a, n_eff_s = res
    beta = 2*math.pi * n_eff_a / wl
    cch = 0.5 * (p["n_ch"] + 1)
    theta = (port - cch) * p["d_io"] / d["fl"]
    dphi = beta * d["dL"] + 2*math.pi * n_eff_s * p["d_aw"] * theta / wl
    c = 0.5 * (p["n_a"] + 1)
    re = im = a_sum = 0.0
    for k in range(1, p["n_a"] + 1):
        phi = (k - c) * dphi
        re += A[k] * math.cos(phi)
        im += A[k] * math.sin(phi)
        a_sum += A[k]
    return (re*re + im*im) / (a_sum * a_sum)


def main(argv):
    if len(argv) < 10:
        print(__doc__)
        sys.exit(1)
    W = float(argv[1]); H = float(argv[2]); delta_pct = float(argv[3])
    lam_c = float(argv[4]); dlam_ch_nm = float(argv[5])
    n_ch = int(argv[6]); n_a = int(argv[7])
    d_aw = float(argv[8]); d_io = float(argv[9])
    m_pref = int(argv[10]) if len(argv) > 10 else 0
    profile = argv[11] if len(argv) > 11 else "gauss"
    w_g = float(argv[12]) if len(argv) > 12 else 0.35
    span = float(argv[13]) if len(argv) > 13 else 1.0
    N = int(argv[14]) if len(argv) > 14 else 900

    dlam_ch_um = dlam_ch_nm * 1e-3
    p = dict(W=W, H=H, delta_pct=delta_pct, lam_c=lam_c, d_aw=d_aw, d_io=d_io,
             n_ch=n_ch, n_a=n_a)
    d = design(W, H, delta_pct, lam_c, dlam_ch_um, n_ch, n_a, d_aw, d_io, m_pref)
    A = coupling_amp(n_a, profile, w_g)

    print(f"n_eff_a = {d['n_eff_a']:.6f}")
    print(f"n_eff_s = {d['n_eff_s']:.6f}")
    print(f"n_g     = {d['n_g']:.6f}")
    print(f"m       = {d['m']}")
    print(f"Delta_L = {d['dL']:.4f} um")
    print(f"FSR     = {d['fsr']*1000:.4f} nm")
    print(f"f_l     = {d['fl']*1e-3:.4f} mm")
    print(f"sweep   = {math.degrees(d['sweep']):.3f} deg")

    print()
    print("# wavelength_um  " + "  ".join(f"port{i}_dB" for i in range(1, n_ch+1)))
    half = 0.5 * span * d["fsr"]
    for n in range(N):
        wl = lam_c - half + 2*half * n / (N - 1)
        row = [f"{wl:.6f}"]
        for i in range(1, n_ch+1):
            T = transmission(wl, i, p, d, A)
            dB = 10*math.log10(T) if T > 1e-12 else -120.0
            row.append(f"{dB:8.3f}")
        print("  ".join(row))


if __name__ == "__main__":
    main(sys.argv)
